### PR TITLE
[WIP] Non-git mixin

### PR DESCRIPTION
**Pull request by Wauplin**
**Disclaimer:** I made quite some mess with the branch/fork so this PR might be temporary one that we close without merging. I created a branch `wauplin-non-git-mixin` from the `main` of the main repo and merged the branch [non-git-mixin](https://github.com/LysandreJik/huggingface_hub/tree/non-git-mixin) from @LysandreJik's fork. My goal was just to start with a clean state once the merge conflicts were resolved. This PR is quite huge because of the new features from the `main` branch. To see what https://github.com/huggingface/huggingface_hub/pull/847 would be after merging see, see the [diff comparison](https://github.com/huggingface/huggingface_hub/compare/main...huggingface:huggingface_hub:wauplin-non-git-mixin?expand=1) (I would rather not start a new PR from this branch to `main` directly just to avoid losing the previous discussion in https://github.com/huggingface/huggingface_hub/pull/847).

**What is done ?**
- resolve conflicts with `main`
- use `upload_folder` in both hub and keras mixins
- some docstring

**What remains ?**
~- `save pretrained` -> signature changes to return `List[str]` instead of `None`. Does it impact other subclasses and how ? Document it (see [comment](https://github.com/huggingface/huggingface_hub/pull/847#discussion_r862016746)).~-> done
~- adapt tests (and create new ones)~ -> done
~- deprecate previous arguments (`repo_path_or_name`, `repo_url`,  `use_auth_token`, `git_user`, `git_email`)~ -> done
~- deprecate `skip_lfs_files` -> introduced in https://github.com/huggingface/huggingface_hub/pull/858 . For what I understand from [this comment](https://github.com/huggingface/huggingface_hub/pull/858#issuecomment-1112906088), `skip_lfs_files` is now useless since we do not pull the repo anyway => to be deprecated~ -> done

**How to deprecate arguments ?**
I am still wondering what is the policy to deprecate things without breaking changes. 

1. In [this change](https://github.com/huggingface/huggingface_hub/pull/847/files#diff-bf673df3e700e1f9ed9cffa6ce51441acd0474bb67f7ac7728aa336d43216285R217), we force the usage of keyword arguments and add the deprecation decorator at the same time. My understanding would be that in this PR/release (0.9 ?) we add the decorator but we still allow it and in next release (0.10 or 0.11 ?) we add the `*`to force the usage of keyword arguments. Do you confirm ?
2. For argument deprecation, is there a utility decorator I could use same as `_deprecate_positional_args` ? I didn't find any but maybe you already have this in another repo ?
3. In some cases, I don't see how we can avoid breaking changes.
    a. Will work: `use_auth_token`. I feel it's fine because we can take the token from either `token` or `use_auth_token` and both old and new behaviour work.
    b. Will not always work: `repo_path_or_name`. What if the user is passing a path to a local directory on its machine ? We cannot handle it with the new behaviour. Does it mean we still need the old implementation with `Repository` as well and make it live until next release ? I am a bit afraid of "dead code" if we do like this. Any opinion ?
    

I am also fine with merging this PR to Lysandre's repo and then continue the discussion in https://github.com/huggingface/huggingface_hub/pull/847

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -1431,7 +1427,7 @@ def list_repo_files(
         )
         return [f.rfilename for f in repo_info.siblings]
 
-    @_deprecate_positional_args
+    @_deprecate_positional_args(version="0.8")
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py]: This was the default value. I either suggest to increase to version to "0.10" or "0.12" or just to remove the deprecation and consider that other libraries made the change already.

IDK the history of this deprecation but my opinion would be to wait for v0.12 and remove all positional args at the same time.

File: src/huggingface_hub/hub_mixin.py [hunk 2]
```diff
@@ -33,44 +34,45 @@ def save_pretrained(
         self,
         save_directory: str,
         config: Optional[dict] = None,
+        # NOTE: without new arguments this behavior (with push_to_hub=True), will be
+        # deprecated as well. Should we already plan to extend the "save_pretrained"
+        # function or should we let this feature die and then force users to do
+        # `model.save_pretrained(...)` and `model.push_to_hub(...)` separately ?
+        # Main argument that is missing I think is "repo_id". All other are optional but
+        # nice to have.
+        # TODO: To be discussed and to remove before merging PR
```
Comment (Wauplin) [inline src/huggingface_hub/hub_mixin.py]: Do not forget to remove.
Previously I guess the other arguments were optional because of the git-based approach.

What is the policy now ? 
Should we add all the arguments here as well (repo_id, branch, create_pr, commit_message, private...) ? Or just `repo_id` which is mandatory to push to hub ? Or just deprecate this feature ?

File: src/huggingface_hub/keras_mixin.py [hunk 3]
```diff
@@ -342,48 +385,115 @@ def push_to_hub_keras(
...
-        exist_ok=True,
-    )
+            if log_dir is not None:
+                # TODO: delete log folder on the hub when it exists
+                # previous implementation using git was doing it in the same git commit
+                # ... to be continued
+                tmp_log_dir = saved_path / "logs"
+                if os.path.exists(tmp_log_dir):
+                    rmtree(tmp_log_dir)
+                copytree(log_dir, tmp_log_dir)
```
Comment (Wauplin) [inline src/huggingface_hub/keras_mixin.py]: Do not forget to remove.

My strategy here would be to use `create_commit` instead of `upload_folder`. This would allow to have a Delete operation along the AddFile operations. Since we are in a temporary directory we can safely list of new files.

I am just not sure if we can delete an entire folder using `create_commit` (and what would happen if we try to delete it and it doesn't exist ? I would prefer to avoid a first HTTP request just to see if logs are there). @LysandreJik an idea ?

File: tests/test_keras_integration.py [hunk 4]
```diff
@@ -440,62 +478,66 @@ def test_override_tensorboard(self):
...
-            )
+            model_info = HfApi(endpoint=ENDPOINT_STAGING).model_info(repo_id)
+            # TODO: this test fails. Both tensorboard.txt and override.txt are in the
+            #       repo at the end.
             self.assertTrue(
                 "logs/override.txt" in [f.rfilename for f in model_info.siblings]
             )
             self.assertFalse(
                 "logs/tensorboard.txt" in [f.rfilename for f in model_info.siblings]
             )
```
Comment (Wauplin) [inline tests/test_keras_integration.py]: Related to my previous message about deleting the logs on the hub while pushing new logs.

File: tests/test_keras_integration.py [hunk 5]
```diff
@@ -440,62 +478,66 @@ def test_override_tensorboard(self):
...
 
-        model_info = HfApi(endpoint=ENDPOINT_STAGING).model_info(
-            f"{USER}/{REPO_NAME}",
-        )
-        self.assertEqual(model_info.modelId, f"{USER}/{REPO_NAME}")
+        model_info = HfApi(endpoint=ENDPOINT_STAGING).model_info(repo_id)
+        self.assertEqual(model_info.modelId, repo_id)
 
-        from_pretrained_keras(f"{WORKING_REPO_DIR}/{REPO_NAME}")
-        self.assertRaises(ValueError, msg="Exception encountered when calling layer*")
```
Comment (Wauplin) [inline tests/test_keras_integration.py:498]: I am not used to unittest (and `self.assertRaises`) but I think this test never worked as intended.
I tried it manually and the `from_pretrained_keras` never fails for me. From [this discussion](https://github.com/huggingface/huggingface_hub/pull/616/files#r796563431), I understand that the goal was to make it fail with a custom layer.

Comment (Wauplin) [reply tests/test_keras_integration.py:498]: Note: I changed it to 
```python
with self.assertRaises(ValueError, msg="Exception encountered when calling layer*" ):
    from_pretrained_keras(tmpdirname)
```
in the tests so now the test fails (no ValueError thrown)

File: src/huggingface_hub/hf_api.py [hunk 6]
```diff
@@ -1431,7 +1428,7 @@ def list_repo_files(
...
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py]: Is this change necessary?

File: src/huggingface_hub/hub_mixin.py (hunk 2 above)
Comment (LysandreJik) [reply src/huggingface_hub/hub_mixin.py]: The signature we have in other libraries seems to be the following (here transformers):
```
            save_directory (`str` or `os.PathLike`):
                Directory where the configuration JSON file will be saved (will be created if it does not exist).
            push_to_hub (`bool`, *optional*, defaults to `False`):
                Whether or not to push your model to the Hugging Face model hub after saving it. You can specify the
                repository you want to push to with `repo_id` (will default to the name of `save_directory` in your
                namespace).
            kwargs:
                Additional key word arguments passed along to the [`~utils.PushToHubMixin.push_to_hub`] method.
```

I think I'd aim at having a similar one here. It seems identical, with a docstring clarification of how the `repo_id` is made:
```
will default to the name of `save_directory` in your namespace
```

File: src/huggingface_hub/hub_mixin.py [hunk 7]
```diff
@@ -214,18 +216,48 @@ def _from_pretrained(
...
+        deprecated_args={
+            "repo_url",
+            "repo_path_or_name",
+            "organization",
+            "use_auth_token",
+            "git_user",
+            "git_email",
+            "skip_lfs_files",
+        },
+    )
```
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py:251]: Love this decorator!

File: src/huggingface_hub/hub_mixin.py [hunk 8]
```diff
@@ -243,50 +275,104 @@ def push_to_hub(
...
+                The git branch on which to push the model. This defaults to
                 the default branch as specified in your repository, which
                 defaults to `"main"`.
+            create_pr (`boolean`, *optional*):
+                Whether or not to create a Pull Request from `branch` with that commit.
+                Defaults to `False`.
             config (`dict`, *optional*):
                 Configuration object to be saved alongside the model weights.
             skip_lfs_files (`bool`, *optional*, defaults to `False`):
                 Whether to skip git-LFS files or not.
```
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py]: As this is now deprecated, I'd remove it from the docstring

File: src/huggingface_hub/hub_mixin.py [hunk 9]
```diff
@@ -243,50 +275,104 @@ def push_to_hub(
...
+                    repo_id=repo_id,
+                    repo_type="model",
+                    path_in_repo="",
+                    token=token,
+                    folder_path=saved_path,
+                    commit_message=commit_message,
+                    revision=branch,
+                    create_pr=create_pr,
+                )
+
```
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py:340]: yes, clean

File: src/huggingface_hub/keras_mixin.py (hunk 3 above)
Comment (LysandreJik) [reply src/huggingface_hub/keras_mixin.py]: I think it could be done with `create_commit` indeed! I'd try a few approaches on a dummy repository to find an approach you're satisfied with. I think a single HTTP call to see what files there are and if there's an existing log dir isn't too expensive so I wouldn't be against having that implemented if you think that's the cleanest.

File: tests/test_keras_integration.py (hunk 5 above)
Comment (LysandreJik) [reply tests/test_keras_integration.py:498]: Ah indeed. Thanks for fixing it!

**Review by LysandreJik — APPROVED** [review]
This looks great! Thanks for working on this, @Wauplin.

I think we can merge this and then ask others for review on the original PR.

File: src/huggingface_hub/hub_mixin.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py]: Makes sense ! I'll adapt like this.

File: src/huggingface_hub/keras_mixin.py (hunk 3 above)
Comment (Wauplin) [reply src/huggingface_hub/keras_mixin.py]: I'll try that indeed. If it's not considered too expensive, I'll indeed do something like that with an extra call.

File: src/huggingface_hub/hf_api.py (hunk 6 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: I just made the `version` keyword required to avoid having a deprecation for which we don't know at which version we want to remove it. Also related to this https://github.com/LysandreJik/huggingface_hub/pull/8#discussion_r938029960 I think we should increase to version 0.12 to deprecate all positional argument at the same time. It doesn't seem we are in a hurry.

**Comment by Wauplin** [comment]
> This looks great! Thanks for working on this, @Wauplin.
> 
> I think we can merge this and then ask others for review on the original PR.

Thanks @LysandreJik for the review. I'll fix the few things you highlighted and then merge it to continue the discussion on the "real" PR :)

File: src/huggingface_hub/hub_mixin.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/hub_mixin.py]: Addressed and tested in [d039ad5](https://github.com/LysandreJik/huggingface_hub/pull/8/commits/d039ad57e14cd1b8e2a8f1e5f5aa23e213f44723).

File: src/huggingface_hub/hub_mixin.py [hunk 10]
```diff
@@ -39,38 +40,59 @@ def save_pretrained(
...
-
-    def _save_pretrained(self, save_directory: str) -> List[str]:
+            kwargs = kwargs.copy()  # soft-copy to avoid mutating input
+            if config is not None:  # kwarg for `push_to_hub`
+                kwargs["config"] = config
+
+            if (
+                # If a deprecated argument is passed, we have to use the deprecated
+                # version of `push_to_hub`.
+                # TODO: remove this possibility in v0.11
```
Comment (Wauplin) [inline src/huggingface_hub/hub_mixin.py]:
```suggestion
                # TODO: remove this possibility in v0.12
```

File: src/huggingface_hub/utils/endpoint_helpers.py [hunk 11]
```diff
@@ -11,13 +11,57 @@
 # limitations under the License.
 """
 Helpful utility functions and classes in relation to exploring API endpoints
-with the aim for a user-friendly interface
+with the aim for a user-friendly interface.
 """
-
+import enum
```
Comment (Wauplin) [inline src/huggingface_hub/utils/endpoint_helpers.py]:
```suggestion
```

File: src/huggingface_hub/keras_mixin.py (hunk 3 above)
Comment (Wauplin) [reply src/huggingface_hub/keras_mixin.py]: Addressed in [bea0b77](https://github.com/LysandreJik/huggingface_hub/pull/8/commits/bea0b775121ff15206c383e293025e5a9d437e09). 

I am not a huge fan of my implementation since it duplicates some code from `upload_folder`. The problem I had is that I want to do exactly what `upload_folder` + add some Delete operations to delete log files in the same commit. I saw 3 solutions here but none were appealing to me:
1. This one. Refactor `upload_folder` and then duplicate 3 lines of code. "It works" but if we patch `upload_folder` in the future, we need to patch the keras mixin as well. 
2. Create 2 commits when using "push to hub": 1 to delete logs, 1 to push the model. I am not a big fan of this one: what is issues between 2 commits ? Can it work with create_pr/custom branch revision ?
3. Add an optional argument `additional_operations` to `upload_folder` to add additional operations to the "upload folder commit". I am not a big fan since I feel it's a quite custom need and `upload_folder` is meant to be a public method.

@LysandreJik If you have any thoughts on how this could be done better, I am interested. Solution 1 is implemented but I am fine with implementing solution 2 or 3 (or any other solution) if you are opinionated about this.

Comment (LysandreJik) [reply src/huggingface_hub/keras_mixin.py]: I think 2 is unideal due to the two commits, and I think it would be quite hard to design an intuitive API for 3. So 1 is fine for me.

I think the changes done in bea0b77 are good but when reading the code it's a bit hard to understand what is the `_generate_url` and why it's used here. I don't understand what is the `content_type` and breaks a bit the flow of code when reading the method.

In that situation I'd either refactor that method to be super clear and understandable, or just keep the copy-paste which is also completely fine with me as it's (imo) more understandable this way and it's only there in a few spots.

Comment (Wauplin) [reply src/huggingface_hub/keras_mixin.py]: Got it. I didn't know how to make it clearer so I guess it's better to just remove it.
I still kept the regex matching logic in a `_parse_revision_from_pr_url` helper just because I'm not a big fan of duplicating parsing code.

**Review by LysandreJik — APPROVED** [review]
This looks great, thanks for iterating @Wauplin!
